File: services/analysis/app/schemas/predictive_maintenance.py
"""
Predictive Maintenance - Pydantic v2 Schemas
API validation schemas for equipment health monitoring and maintenance tracking
"""

from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator
from enum import Enum


# ============================================================================
# Enums
# ============================================================================

class EquipmentStatus(str, Enum):
    """Equipment health status"""
    HEALTHY = "healthy"
    WARNING = "warning"
    CRITICAL = "critical"


class MaintenanceEventType(str, Enum):
    """Type of maintenance event"""
    PREVENTIVE = "preventive"
    CORRECTIVE = "corrective"
    PREDICTIVE = "predictive"
    INSPECTION = "inspection"


# ============================================================================
# Equipment Health Schemas
# ============================================================================

class EquipmentHealthBase(BaseModel):
    """Base schema for Equipment Health"""
    equipment_id: UUID = Field(..., description="Equipment UUID")
    equipment_name: str = Field(..., min_length=1, max_length=255, description="Equipment name")
    equipment_type: Optional[str] = Field(None, max_length=100, description="Equipment type/category")

    # Health metrics
    health_score: float = Field(..., ge=0.0, le=100.0, description="Health score (0-100)")
    predicted_failure_date: Optional[datetime] = Field(
        None,
        description="Predicted failure date based on analysis"
    )
    confidence: float = Field(..., ge=0.0, le=1.0, description="Prediction confidence (0-1)")
    failure_probability: float = Field(..., ge=0.0, le=1.0, description="Failure probability (0-1)")

    # Reliability metrics
    mtbf_hours: Optional[float] = Field(None, ge=0.0, description="Mean Time Between Failures (hours)")
    mttr_hours: Optional[float] = Field(None, ge=0.0, description="Mean Time To Repair (hours)")

    # Maintenance schedule
    last_maintenance_date: Optional[datetime] = Field(None, description="Last maintenance date")
    next_maintenance_date: Optional[datetime] = Field(None, description="Next scheduled maintenance date")

    # Data and analysis
    sensor_data: Optional[Dict[str, Any]] = Field(
        None,
        description="Latest sensor readings",
        examples=[{
            "temperature": 75.2,
            "vibration": 0.05,
            "pressure": 101.3,
            "runtime_hours": 1250
        }]
    )
    anomalies: Optional[Dict[str, Any]] = Field(
        None,
        description="Detected anomalies",
        examples=[{
            "detected": ["vibration_spike", "temperature_variance"],
            "severity": "medium",
            "timestamp": "2025-11-13T10:30:00Z"
        }]
    )
    recommendations: Optional[Dict[str, Any]] = Field(
        None,
        description="Maintenance recommendations",
        examples=[{
            "actions": ["inspect_bearings", "check_lubrication"],
            "priority": "high",
            "estimated_cost": 500.0
        }]
    )

    @field_validator('health_score')
    @classmethod
    def validate_health_score(cls, v: float) -> float:
        """Validate health score is within range"""
        if not 0.0 <= v <= 100.0:
            raise ValueError("Health score must be between 0 and 100")
        return round(v, 2)

    @field_validator('confidence', 'failure_probability')
    @classmethod
    def validate_probability(cls, v: float) -> float:
        """Validate probability values are within range"""
        if not 0.0 <= v <= 1.0:
            raise ValueError("Probability values must be between 0 and 1")
        return round(v, 4)


class EquipmentHealthCreate(EquipmentHealthBase):
    """Schema for creating an Equipment Health record"""
    org_id: UUID = Field(..., description="Organization UUID")


class EquipmentHealthUpdate(BaseModel):
    """Schema for updating an Equipment Health record"""
    equipment_name: Optional[str] = Field(None, min_length=1, max_length=255)
    equipment_type: Optional[str] = Field(None, max_length=100)
    health_score: Optional[float] = Field(None, ge=0.0, le=100.0)
    predicted_failure_date: Optional[datetime] = None
    confidence: Optional[float] = Field(None, ge=0.0, le=1.0)
    failure_probability: Optional[float] = Field(None, ge=0.0, le=1.0)
    mtbf_hours: Optional[float] = Field(None, ge=0.0)
    mttr_hours: Optional[float] = Field(None, ge=0.0)
    last_maintenance_date: Optional[datetime] = None
    next_maintenance_date: Optional[datetime] = None
    sensor_data: Optional[Dict[str, Any]] = None
    anomalies: Optional[Dict[str, Any]] = None
    recommendations: Optional[Dict[str, Any]] = None


class EquipmentHealthSchema(EquipmentHealthBase):
    """Schema for Equipment Health response"""
    id: UUID
    org_id: UUID
    status: EquipmentStatus = Field(..., description="Calculated equipment status")
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


# ============================================================================
# Maintenance Event Schemas
# ============================================================================

class MaintenanceEventBase(BaseModel):
    """Base schema for Maintenance Event"""
    equipment_id: UUID = Field(..., description="Equipment UUID")
    event_type: MaintenanceEventType = Field(..., description="Type of maintenance event")
    performed_date: datetime = Field(..., description="Date maintenance was performed")
    downtime_hours: float = Field(default=0.0, ge=0.0, description="Equipment downtime in hours")
    cost: float = Field(default=0.0, ge=0.0, description="Maintenance cost")

    # Personnel and details
    technician: Optional[str] = Field(None, max_length=255, description="Technician or team name")
    description: Optional[str] = Field(None, description="Detailed description of work performed")
    parts_replaced: Optional[Dict[str, Any]] = Field(
        None,
        description="Parts replaced during maintenance",
        examples=[{
            "parts": [
                {"name": "bearing", "quantity": 2, "part_number": "BRG-123"},
                {"name": "seal", "quantity": 1, "part_number": "SEAL-456"}
            ]
        }]
    )
    notes: Optional[str] = Field(None, description="Additional notes or observations")

    @field_validator('downtime_hours', 'cost')
    @classmethod
    def validate_non_negative(cls, v: float) -> float:
        """Validate values are non-negative"""
        if v < 0:
            raise ValueError("Value must be non-negative")
        return round(v, 2)


class MaintenanceEventCreate(MaintenanceEventBase):
    """Schema for creating a Maintenance Event"""
    org_id: UUID = Field(..., description="Organization UUID")


class MaintenanceEventUpdate(BaseModel):
    """Schema for updating a Maintenance Event"""
    event_type: Optional[MaintenanceEventType] = None
    performed_date: Optional[datetime] = None
    downtime_hours: Optional[float] = Field(None, ge=0.0)
    cost: Optional[float] = Field(None, ge=0.0)
    technician: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = None
    parts_replaced: Optional[Dict[str, Any]] = None
    notes: Optional[str] = None


class MaintenanceEventSchema(MaintenanceEventBase):
    """Schema for Maintenance Event response"""
    id: UUID
    org_id: UUID
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ============================================================================
# Dashboard Summary Schema
# ============================================================================

class PredictiveMaintenanceDashboard(BaseModel):
    """Schema for predictive maintenance dashboard statistics"""
    total_equipment: int = Field(..., description="Total number of equipment being monitored")
    healthy_count: int = Field(..., description="Number of equipment in healthy status")
    warning_count: int = Field(..., description="Number of equipment in warning status")
    critical_count: int = Field(..., description="Number of equipment in critical status")
    avg_health_score: float = Field(..., description="Average health score across all equipment")
    predictions_this_month: int = Field(
        ...,
        description="Number of failure predictions for this month"
    )
    total_downtime_hours: float = Field(
        ...,
        description="Total equipment downtime hours (last 30 days)"
    )
    total_maintenance_cost: float = Field(
        ...,
        description="Total maintenance cost (last 30 days)"
    )

    # Breakdowns
    by_status: Dict[str, int] = Field(
        default_factory=dict,
        description="Equipment count by status"
    )
    by_equipment_type: Dict[str, int] = Field(
        default_factory=dict,
        description="Equipment count by type"
    )


# ============================================================================
# Prediction Schema
# ============================================================================

class FailurePrediction(BaseModel):
    """Schema for failure prediction details"""
    equipment_id: UUID
    equipment_name: str
    equipment_type: Optional[str] = None
    health_score: float
    predicted_failure_date: datetime
    days_until_failure: int
    failure_probability: float
    confidence: float
    status: EquipmentStatus
    recommendations: Optional[Dict[str, Any]] = None

    model_config = ConfigDict(from_attributes=True)


# ============================================================================
# Query Schemas
# ============================================================================

class EquipmentHealthQuery(BaseModel):
    """Schema for querying equipment health records"""
    org_id: Optional[UUID] = None
    equipment_id: Optional[UUID] = None
    status: Optional[EquipmentStatus] = None
    equipment_type: Optional[str] = None

    # Health score range
    min_health_score: Optional[float] = Field(None, ge=0.0, le=100.0)
    max_health_score: Optional[float] = Field(None, ge=0.0, le=100.0)

    # Pagination
    skip: int = Field(default=0, ge=0)
    limit: int = Field(default=100, ge=1, le=1000)

    # Sorting
    sort_by: str = Field(default="health_score", description="Field to sort by")
    sort_desc: bool = Field(default=True, description="Sort descending")


class MaintenanceEventQuery(BaseModel):
    """Schema for querying maintenance events"""
    org_id: Optional[UUID] = None
    equipment_id: Optional[UUID] = None
    event_type: Optional[MaintenanceEventType] = None

    # Date range filters
    performed_date_from: Optional[datetime] = None
    performed_date_to: Optional[datetime] = None

    # Pagination
    skip: int = Field(default=0, ge=0)
    limit: int = Field(default=100, ge=1, le=1000)

    # Sorting
    sort_by: str = Field(default="performed_date", description="Field to sort by")
    sort_desc: bool = Field(default=True, description="Sort descending")

File: apps/web/src/lib/api/predictive-maintenance.ts
/**
 * Predictive Maintenance API Client
 * Equipment health monitoring, failure prediction, and maintenance tracking
 */

const API_BASE_URL = process.env.NEXT_PUBLIC_ANALYSIS_API_URL || 'http://localhost:8001'

// Enums
export type EquipmentStatus = 'healthy' | 'warning' | 'critical'
export type MaintenanceEventType = 'preventive' | 'corrective' | 'predictive' | 'inspection'

// Equipment Health interfaces
export interface EquipmentHealth {
  id: string
  org_id: string
  equipment_id: string
  equipment_name: string
  equipment_type?: string

  // Health metrics
  health_score: number
  predicted_failure_date?: string
  confidence: number
  failure_probability: number

  // Reliability metrics
  mtbf_hours?: number
  mttr_hours?: number

  // Maintenance schedule
  last_maintenance_date?: string
  next_maintenance_date?: string

  // Data and analysis
  sensor_data?: Record<string, any>
  anomalies?: Record<string, any>
  recommendations?: Record<string, any>

  // Status
  status: EquipmentStatus
  created_at: string
  updated_at?: string
}

export interface EquipmentHealthCreate {
  org_id: string
  equipment_id: string
  equipment_name: string
  equipment_type?: string
  health_score: number
  predicted_failure_date?: string
  confidence: number
  failure_probability: number
  mtbf_hours?: number
  mttr_hours?: number
  last_maintenance_date?: string
  next_maintenance_date?: string
  sensor_data?: Record<string, any>
  anomalies?: Record<string, any>
  recommendations?: Record<string, any>
}

export interface EquipmentHealthUpdate {
  equipment_name?: string
  equipment_type?: string
  health_score?: number
  predicted_failure_date?: string
  confidence?: number
  failure_probability?: number
  mtbf_hours?: number
  mttr_hours?: number
  last_maintenance_date?: string
  next_maintenance_date?: string
  sensor_data?: Record<string, any>
  anomalies?: Record<string, any>
  recommendations?: Record<string, any>
}

// Maintenance Event interfaces
export interface MaintenanceEvent {
  id: string
  org_id: string
  equipment_id: string
  event_type: MaintenanceEventType
  performed_date: string
  downtime_hours: number
  cost: number
  technician?: string
  description?: string
  parts_replaced?: Record<string, any>
  notes?: string
  created_at: string
}

export interface MaintenanceEventCreate {
  org_id: string
  equipment_id: string
  event_type: MaintenanceEventType
  performed_date: string
  downtime_hours?: number
  cost?: number
  technician?: string
  description?: string
  parts_replaced?: Record<string, any>
  notes?: string
}

export interface MaintenanceEventUpdate {
  event_type?: MaintenanceEventType
  performed_date?: string
  downtime_hours?: number
  cost?: number
  technician?: string
  description?: string
  parts_replaced?: Record<string, any>
  notes?: string
}

// Dashboard interface
export interface PredictiveMaintenanceDashboard {
  total_equipment: number
  healthy_count: number
  warning_count: number
  critical_count: number
  avg_health_score: number
  predictions_this_month: number
  total_downtime_hours: number
  total_maintenance_cost: number
  by_status: Record<string, number>
  by_equipment_type: Record<string, number>
}

// Failure Prediction interface
export interface FailurePrediction {
  equipment_id: string
  equipment_name: string
  equipment_type?: string
  health_score: number
  predicted_failure_date: string
  days_until_failure: number
  failure_probability: number
  confidence: number
  status: EquipmentStatus
  recommendations?: Record<string, any>
}

class PredictiveMaintenanceAPI {
  private baseUrl: string

  constructor(baseUrl: string = API_BASE_URL) {
    this.baseUrl = baseUrl
  }

  // ============================================================================
  // Equipment Health Records
  // ============================================================================

  // Get all equipment health records
  async getHealthRecords(params?: {
    org_id?: string
    status?: EquipmentStatus
    equipment_type?: string
    equipment_id?: string
    min_health_score?: number
    max_health_score?: number
    skip?: number
    limit?: number
    sort_by?: string
    sort_desc?: boolean
  }): Promise<EquipmentHealth[]> {
    const queryParams = new URLSearchParams()
    if (params?.org_id) queryParams.append('org_id', params.org_id)
    if (params?.status) queryParams.append('status', params.status)
    if (params?.equipment_type) queryParams.append('equipment_type', params.equipment_type)
    if (params?.equipment_id) queryParams.append('equipment_id', params.equipment_id)
    if (params?.min_health_score !== undefined) queryParams.append('min_health_score', params.min_health_score.toString())
    if (params?.max_health_score !== undefined) queryParams.append('max_health_score', params.max_health_score.toString())
    if (params?.skip !== undefined) queryParams.append('skip', params.skip.toString())
    if (params?.limit) queryParams.append('limit', params.limit.toString())
    if (params?.sort_by) queryParams.append('sort_by', params.sort_by)
    if (params?.sort_desc !== undefined) queryParams.append('sort_desc', params.sort_desc.toString())

    const response = await fetch(`${this.baseUrl}/api/v1/predictive-maintenance/health?${queryParams}`)
    if (!response.ok) {
      throw new Error(`Failed to fetch health records: ${response.statusText}`)
    }
    return response.json()
  }

  // Get specific health record
  async getHealthRecord(id: string): Promise<EquipmentHealth> {
    const response = await fetch(`${this.baseUrl}/api/v1/predictive-maintenance/health/${id}`)
    if (!response.ok) {
      throw new Error(`Failed to fetch health record: ${response.statusText}`)
    }
    return response.json()
  }

  // Create health record
  async createHealthRecord(data: EquipmentHealthCreate): Promise<EquipmentHealth> {
    const response = await fetch(`${this.baseUrl}/api/v1/predictive-maintenance/health`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    })
    if (!response.ok) {
      throw new Error(`Failed to create health record: ${response.statusText}`)
    }
    return response.json()
  }

  // Update health record
  async updateHealthRecord(id: string, data: EquipmentHealthUpdate): Promise<EquipmentHealth> {
    const response = await fetch(`${this.baseUrl}/api/v1/predictive-maintenance/health/${id}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    })
    if (!response.ok) {
      throw new Error(`Failed to update health record: ${response.statusText}`)
    }
    return response.json()
  }

  // Get latest health for specific equipment
  async getLatestEquipmentHealth(equipmentId: string): Promise<EquipmentHealth> {
    const response = await fetch(`${this.baseUrl}/api/v1/predictive-maintenance/equipment/${equipmentId}/health`)
    if (!response.ok) {
      throw new Error(`Failed to fetch equipment health: ${response.statusText}`)
    }
    return response.json()
  }

  // ============================================================================
  // Maintenance Events
  // ============================================================================

  // Get all maintenance events
  async getMaintenanceEvents(params?: {
    org_id?: string
    equipment_id?: string
    event_type?: MaintenanceEventType
    performed_date_from?: string
    performed_date_to?: string
    skip?: number
    limit?: number
    sort_by?: string
    sort_desc?: boolean
  }): Promise<MaintenanceEvent[]> {
    const queryParams = new URLSearchParams()
    if (params?.org_id) queryParams.append('org_id', params.org_id)
    if (params?.equipment_id) queryParams.append('equipment_id', params.equipment_id)
    if (params?.event_type) queryParams.append('event_type', params.event_type)
    if (params?.performed_date_from) queryParams.append('performed_date_from', params.performed_date_from)
    if (params?.performed_date_to) queryParams.append('performed_date_to', params.performed_date_to)
    if (params?.skip !== undefined) queryParams.append('skip', params.skip.toString())
    if (params?.limit) queryParams.append('limit', params.limit.toString())
    if (params?.sort_by) queryParams.append('sort_by', params.sort_by)
    if (params?.sort_desc !== undefined) queryParams.append('sort_desc', params.sort_desc.toString())

    const response = await fetch(`${this.baseUrl}/api/v1/predictive-maintenance/events?${queryParams}`)
    if (!response.ok) {
      throw new Error(`Failed to fetch maintenance events: ${response.statusText}`)
    }
    return response.json()
  }

  // Create maintenance event
  async createMaintenanceEvent(data: MaintenanceEventCreate): Promise<MaintenanceEvent> {
    const response = await fetch(`${this.baseUrl}/api/v1/predictive-maintenance/events`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    })
    if (!response.ok) {
      throw new Error(`Failed to create maintenance event: ${response.statusText}`)
    }
    return response.json()
  }

  // ============================================================================
  // Dashboard & Analytics
  // ============================================================================

  // Get dashboard statistics
  async getDashboard(params?: { org_id?: string }): Promise<PredictiveMaintenanceDashboard> {
    const queryParams = new URLSearchParams()
    if (params?.org_id) queryParams.append('org_id', params.org_id)

    const response = await fetch(`${this.baseUrl}/api/v1/predictive-maintenance/dashboard?${queryParams}`)
    if (!response.ok) {
      throw new Error(`Failed to fetch dashboard: ${response.statusText}`)
    }
    return response.json()
  }

  // Get failure predictions
  async getFailurePredictions(params?: {
    org_id?: string
    days_ahead?: number
    min_probability?: number
    skip?: number
    limit?: number
  }): Promise<FailurePrediction[]> {
    const queryParams = new URLSearchParams()
    if (params?.org_id) queryParams.append('org_id', params.org_id)
    if (params?.days_ahead) queryParams.append('days_ahead', params.days_ahead.toString())
    if (params?.min_probability !== undefined) queryParams.append('min_probability', params.min_probability.toString())
    if (params?.skip !== undefined) queryParams.append('skip', params.skip.toString())
    if (params?.limit) queryParams.append('limit', params.limit.toString())

    const response = await fetch(`${this.baseUrl}/api/v1/predictive-maintenance/predictions?${queryParams}`)
    if (!response.ok) {
      throw new Error(`Failed to fetch failure predictions: ${response.statusText}`)
    }
    return response.json()
  }

  // ============================================================================
  // Helper Methods
  // ============================================================================

  getHealthyEquipment(healthRecords: EquipmentHealth[]) {
    return healthRecords.filter(h => h.status === 'healthy')
  }

  getWarningEquipment(healthRecords: EquipmentHealth[]) {
    return healthRecords.filter(h => h.status === 'warning')
  }

  getCriticalEquipment(healthRecords: EquipmentHealth[]) {
    return healthRecords.filter(h => h.status === 'critical')
  }

  getEventsByType(events: MaintenanceEvent[]) {
    return events.reduce((acc, e) => {
      if (!acc[e.event_type]) {
        acc[e.event_type] = []
      }
      acc[e.event_type].push(e)
      return acc
    }, {} as Record<MaintenanceEventType, MaintenanceEvent[]>)
  }

  calculateTotalDowntime(events: MaintenanceEvent[]): number {
    return events.reduce((sum, e) => sum + e.downtime_hours, 0)
  }

  calculateTotalCost(events: MaintenanceEvent[]): number {
    return events.reduce((sum, e) => sum + e.cost, 0)
  }
}

export const predictiveMaintenanceApi = new PredictiveMaintenanceAPI()
export default predictiveMaintenanceApi
